frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx

```diff
@@ -173,167 +174,230 @@ interface DashboardProps {
   ) => void;
 }
 
-interface DashboardState {
-  error: unknown;
-  parametersListLength: number;
-  hasScroll: boolean;
-}
-
-class DashboardInner extends Component<DashboardProps, DashboardState> {
-  state = {
-    error: null,
-    parametersListLength: 0,
-    hasScroll: getMainElement()?.scrollTop > 0,
-  };
-
-  static defaultProps = {
-    isSharing: false,
-  };
-
-  static getDerivedStateFromProps(
-    { parameters }: DashboardProps,
-    { parametersListLength }: DashboardState,
-  ) {
-    const visibleParameters = getVisibleParameters(parameters);
-    return visibleParameters.length !== parametersListLength
-      ? { parametersListLength: visibleParameters.length }
-      : null;
-  }
-
-  async componentDidMount() {
-    await this.loadDashboard(this.props.dashboardId);
-    getMainElement().addEventListener("scroll", this.onMainScroll, {
-      capture: false,
-      passive: true,
-    });
-  }
-
-  async componentDidUpdate(prevProps: DashboardProps) {
-    if (prevProps.dashboardId !== this.props.dashboardId) {
-      await this.loadDashboard(this.props.dashboardId);
-      return;
-    }
-
-    if (!_.isEqual(prevProps.selectedTabId, this.props.selectedTabId)) {
-      this.props.fetchDashboardCardData();
-      this.props.fetchDashboardCardMetadata();
-      return;
+function DashboardInner(props: DashboardProps) {
+  const {
+    addCardOnLoad,
+    addCardToDashboard,
+    addParameter,
+    cancelFetchDashboardCardData,
+    closeNavbar,
+    dashboard,
+    dashboardId,
+    draftParameterValues,
+    editingOnLoad,
+    editingParameter,
+    fetchDashboard,
+    fetchDashboardCardData,
+    fetchDashboardCardMetadata,
+    initialize,
+    isAutoApplyFilters,
+    isEditing,
+    isFullscreen,
+    isHeaderVisible,
+    isNavigatingBackToDashboard,
+    isNightMode,
+    isSharing,
+    loadDashboardParams,
+    location,
+    onRefreshPeriodChange,
+    parameterValues,
+    parameters,
+    selectedTabId,
+    setDashboardAttributes,
+    setEditingDashboard,
+    setEditingParameter,
+    setErrorPage,
+    setParameterIndex,
+    setParameterValue,
+    setSharing,
+    toggleSidebar,
+  } = props;
+
+  const [isInitialized, setIsInitialized] = useState(false);
+  const [error, setError] = useState(null);
+  const [hasScroll, setHasScroll] = useState(getMainElement()?.scrollTop > 0);
+
+  const previousDashboard = usePrevious(dashboard);
+  const previousDashboardId = usePrevious(dashboardId);
+  const previousTabId = usePrevious(selectedTabId);
+  const previousParameterValues = usePrevious(parameterValues);
+
+  const visibleParameters = useMemo(
+    () => getVisibleParameters(parameters),
+    [parameters],
+  );
+
+  const tabHasCards = useMemo(() => {
+    if (!Array.isArray(dashboard?.dashcards)) {
+      return false;
     }
+    const tabDashCards = dashboard.dashcards.filter(
+      dc => dc.dashboard_tab_id === selectedTabId,
+    );
```

Comment:
Previously it also checked that `selectedTabId` is not `undefined`. And `dc.dashboard_tab_id` can be `undefined`.